### ISSUE TITLE
Module validation reports incorrect cfn-lint warnings 

### ISSUE THREAD

**Issue by michaelwittig** [aws-cloudformation/cloudformation-cli#689]
I have the following fragment (`fragments/module.yml`):
```yaml
---
AWSTemplateFormatVersion: '2010-09-09'
Description: 'VPC Subnet'
Parameters:
  VpcId:
    Description: 'VPC id'
    Type: String
  VpcCidrBlock:
    Description: 'VPC CIDR block'
    Type: String
  InternetGatewayId:
    Description: 'VPC Internet Gateway id'
    Type: String
  Reach:
    Description: 'Should the subnet has a route to the Internet?'
    Type: String
    AllowedValues: [private, public]
  AZIndex:
    Description: 'Index of the AZ where the subnet should be created in'
    Type: Number
    MinValue: 0
    MaxValue: 2
  AZChar:
    Description: 'Availability zone char'
    Type: String
    AllowedValues: [A, B, C]
  SubnetIndex:
    Description: 'Index of the subnet'
    Type: Number
    MinValue: 0
    MaxValue: 5
  SubnetCount:
    Description: 'To slice the IP address ranges you need to specify how many subnets you want to create in the VPC'
    Type: Number
    MinValue: 1
    MaxValue: 6
Conditions:
  IsPublic: !Equals [!Ref Reach, 'public']
Resources:
  Subnet:
    Type: 'AWS::EC2::Subnet'
    Properties:
      AvailabilityZone: !Select [!Ref AZIndex, !GetAZs '']
      CidrBlock: !Select [!Ref SubnetIndex, !Cidr [!Ref VpcCidrBlock, !Ref SubnetCount, 12]]
      VpcId: !Ref VpcId
      Tags:
      - Key: Name
        Value: !Sub '${AZChar} ${Reach}'
      - Key: Reach
        Value: !Ref Reach
  RouteTable:
    Type: 'AWS::EC2::RouteTable'
    Properties:
      VpcId: !Ref VpcId
      Tags:
      - Key: Name
        Value: !Sub '${AZChar} ${Reach}'
  SubnetRouteTableAssociation:
    Type: 'AWS::EC2::SubnetRouteTableAssociation'
    Properties:
      SubnetId: !Ref Subnet
      RouteTableId: !Ref RouteTable
  RouteInternet:
    Type: 'AWS::EC2::Route'
    Condition: IsPublic
    Properties:
      RouteTableId: !Ref RouteTable
      DestinationCidrBlock: '0.0.0.0/0'
      GatewayId: !Ref InternetGatewayId
  NetworkAcl:
    Type: 'AWS::EC2::NetworkAcl'
    Properties:
      VpcId: !Ref VpcId
      Tags:
      - Key: Name
        Value: !Sub '${AZChar} ${Reach}'
  NetworkAclEntryAllowAllInbound:
    Type: 'AWS::EC2::NetworkAclEntry'
    Properties:
      NetworkAclId: !Ref NetworkAcl
      RuleNumber: 99
      Protocol: -1
      RuleAction: allow
      Egress: false
      CidrBlock: '0.0.0.0/0'
  NetworkAclEntryAllowAllOutbound:
    Type: 'AWS::EC2::NetworkAclEntry'
    Properties:
      NetworkAclId: !Ref NetworkAcl
      RuleNumber: 99
      Protocol: -1
      RuleAction: allow
      Egress: true
      CidrBlock: '0.0.0.0/0'
  SubnetNetworkAclAssociation:
    Type: 'AWS::EC2::SubnetNetworkAclAssociation'
    Properties:
      SubnetId: !Ref Subnet
      NetworkAclId: !Ref NetworkAcl
Outputs:
  Id:
    Value: !Ref Subnet
  RouteTableId:
    Value: !Ref RouteTable
  AvailabilityZone:
    Value: !Select [!Ref AZIndex, !GetAZs '']

```

When I run the `cfn validate` command I get a ton of warnings from `cfn-lint`:

```
$ cfn --version
cfn 0.2.3
$ cfn validate
Module fragment is probably valid, but there are warnings/errors from cfn-lint (https://github.com/aws-cloudformation/cfn-python-lint):
	Parameter VpcId not used. (from rule W2001: Check if Parameters are Used)
	Parameter VpcCidrBlock not used. (from rule W2001: Check if Parameters are Used)
	Parameter InternetGatewayId not used. (from rule W2001: Check if Parameters are Used)
	Parameter Reach not used. (from rule W2001: Check if Parameters are Used)
	Parameter AZIndex not used. (from rule W2001: Check if Parameters are Used)
	Parameter AZChar not used. (from rule W2001: Check if Parameters are Used)
	Parameter SubnetIndex not used. (from rule W2001: Check if Parameters are Used)
	Parameter SubnetCount not used. (from rule W2001: Check if Parameters are Used)
	Condition IsPublic has invalid property (from rule E8001: Conditions have appropriate properties)
	You must specify a valid value for AvailabilityZone (Select [Ref AZIndex, GetAZs '']). Valid values are ["af-south-1a", "af-south-1b", "af-south-1c", "ap-east-1a", "ap-east-1b", "ap-east-1c", "ap-northeast-1a", "ap-northeast-1b", "ap-northeast-1c", "ap-northeast-1d", "ap-northeast-2a", "ap-northeast-2b", "ap-northeast-2c", "ap-northeast-2d", "ap-northeast-3a", "ap-south-1a", "ap-south-1b", "ap-south-1c", "ap-southeast-1a", "ap-southeast-1b", "ap-southeast-1c", "ap-southeast-2a", "ap-southeast-2b", "ap-southeast-2c", "ca-central-1a", "ca-central-1b", "ca-central-1d", "cn-north-1a", "cn-north-1b", "cn-northwest-1a", "cn-northwest-1b", "cn-northwest-1c", "eu-central-1a", "eu-central-1b", "eu-central-1c", "eu-north-1a", "eu-north-1b", "eu-north-1c", "eu-south-1a", "eu-south-1b", "eu-south-1c", "eu-west-1a", "eu-west-1b", "eu-west-1c", "eu-west-2a", "eu-west-2b", "eu-west-2c", "eu-west-3a", "eu-west-3b", "eu-west-3c", "me-south-1a", "me-south-1b", "me-south-1c", "sa-east-1a", "sa-east-1b", "sa-east-1c", "us-east-1a", "us-east-1b", "us-east-1c", "us-east-1d", "us-east-1e", "us-east-1f", "us-east-2a", "us-east-2b", "us-east-2c", "us-gov-east-1a", "us-gov-east-1b", "us-gov-east-1c", "us-gov-west-1a", "us-gov-west-1b", "us-gov-west-1c", "us-west-1a", "us-west-1b", "us-west-1c", "us-west-2a", "us-west-2b", "us-west-2c", "us-west-2d", "us-west-2-lax-1a", "us-west-2-lax-1b"] (from rule E3030: Check if properties have a valid value)
	Don't hardcode Select [Ref AZIndex, GetAZs ''] for AvailabilityZones (from rule W3010: Availability Zone Parameters should not be hardcoded)
	CidrBlock contains invalid characters (Pattern: x.x.x.x/y) at Resources/Subnet/Properties/CidrBlock (from rule E3031: Check if property values adhere to a specific pattern)
	Found an embedded parameter "${AZChar}" outside of an "Fn::Sub" at Resources/Subnet/Properties/Tags/0/Value (from rule E1029: Sub is required if a variable is used in a string)
	Found an embedded parameter "${Reach}" outside of an "Fn::Sub" at Resources/Subnet/Properties/Tags/0/Value (from rule E1029: Sub is required if a variable is used in a string)
	Found an embedded parameter "${AZChar}" outside of an "Fn::Sub" at Resources/RouteTable/Properties/Tags/0/Value (from rule E1029: Sub is required if a variable is used in a string)
	Found an embedded parameter "${Reach}" outside of an "Fn::Sub" at Resources/RouteTable/Properties/Tags/0/Value (from rule E1029: Sub is required if a variable is used in a string)
	Found an embedded parameter "${AZChar}" outside of an "Fn::Sub" at Resources/NetworkAcl/Properties/Tags/0/Value (from rule E1029: Sub is required if a variable is used in a string)
	Found an embedded parameter "${Reach}" outside of an "Fn::Sub" at Resources/NetworkAcl/Properties/Tags/0/Value (from rule E1029: Sub is required if a variable is used in a string)
```

Running `cfn-lint` against the template works fine:
```
$ cfn-lint -v
cfn-lint 0.45.0
$ cfn-lint -t fragments/module.yml 
```

Why does `cfn` produce all those warnings`

**Comment by kddejong** [comment]
Looks like the ! before function is being removed.  This was probably not as much of an issue when we were parsing just json.

https://github.com/aws-cloudformation/cloudformation-cli/blob/b01caaea2a76416c4f2108efd4f165536b9d52a0/src/rpdk/core/fragment/generator.py#L388

https://github.com/aws-cloudformation/cloudformation-cli/blob/b01caaea2a76416c4f2108efd4f165536b9d52a0/src/rpdk/core/fragment/generator.py#L107-L108

by the time its passed to cfn-lint `!Select [!Ref AZIndex, !GetAZs '']` becomes `Select [Ref AZIndex, GetAZs '']`.  We may need to switch to using the original template not the one use to build the schema.

**Comment by MalikAtalla-AWS** [comment]
Will be resolved in versions >= 0.2.6
